pkg/ottl/functions.go

```diff
@@ -18,11 +18,38 @@ import (
 	"errors"
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
 
+	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
+// Arguments holds the arguments for an OTTL function, with arguments
+// specified as fields on a struct. Argument ordering is defined
+type Arguments interface{}
+
+// FunctionContext contains data provided by the Collector
+// component to the OTTL for use in functions.
+type FunctionContext struct {
+	Set component.TelemetrySettings
+}
+
+// Factory defines an OTTL function factory that will generate an OTTL
+// function to be called based on an invocation within a statement.
+type Factory[K any] interface {
```

Comment:
Follow the pattern in the core, add a private func and a way to construct instances, allows to extend in the future.

Reply:
Right now we can extend factories by providing optional interfaces for them to implement. Do you think the component factory pattern of using options to allow extending functionality offers anything that we won't be able to do with the current approach?

Reply:
I tried this out and was able to get a workable solution, though I haven't thought through all the implications.

```golang
type Expr[K any] interface {
	Eval(ctx context.Context, tCtx K) (any, error)
	PutArgs(fCtx ottl.FunctionContext, oArgs ottl.Arguments) error
}

type baseExpr[K any, A any] struct {
	args *A
	fCtx *ottl.FunctionContext
}

func (d *baseExpr[K, A]) PutArgs(fCtx ottl.FunctionContext, oArgs ottl.Arguments) error {
	args, ok := oArgs.(*A)

	if !ok {
		return fmt.Errorf("Arguments are the wrong type. This is a bug in the OTTL.")
	}

	d.args = args
	d.fCtx = &fCtx

	return nil
}

type functionFactory[K any, A any] struct {
	name string
	args A
	expr Expr[K]
}

func (f functionFactory[K, A]) Name() string { return f.name }

func (f functionFactory[K, A]) CreateEmptyArguments() ottl.Arguments {
	return f.args
}

func (f functionFactory[K, A]) CreateFunction(fCtx ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
	err := f.expr.PutArgs(fCtx, oArgs)

	if err != nil {
		return nil, err
	}

	return f.expr.Eval, nil

}

func createFactory[K any, A any](name string, args A, expr Expr[K]) ottl.Factory[K] {
	return functionFactory[K, A]{
		name: name,
		args: args,
		expr: expr,
	}
}

func NewDeleteKeyFactory[K any]() ottl.Factory[K] {
	return createFactory[K]("delete_key", &DeleteKeyArguments[K]{}, &deleteKeyExpr[K]{})
}

type deleteKeyExpr[K any] struct {
	baseExpr[K, DeleteKeyArguments[K]]
}

func (d *deleteKeyExpr[K]) Eval(ctx context.Context, tCtx K) (any, error) {
	val, err := d.args.Target.Get(ctx, tCtx)
	if err != nil {
		return nil, err
	}
	val.Remove(d.args.Key)
	return nil, nil
}
```

Reply:
This is interesting, but feels pretty complicated. I think it would feel a lot better if we didn't need the `Expr` interface.  Is there any way we can drop it?

```go
// Arguments holds the arguments for an OTTL function, with arguments
// specified as fields on a struct. Argument ordering is defined
type Arguments interface{}

// FunctionContext contains data provided by the Collector
// component to the OTTL for use in functions.
type FunctionContext struct {
	Set component.TelemetrySettings
}

// Factory defines an OTTL function factory that will generate an OTTL
// function to be called based on an invocation within a statement.
type Factory[K any] interface {
	// Name is the canonical name to be used by the user when invocating
	// the function generated by this Factory.
	Name() string

	// CreateEmptyArguments initializes an Arguments struct specific to this
	// Factory containing the arguments for the function.
	CreateEmptyArguments() Arguments

	// CreateFunction creates an OTTL function that will use the given Arguments.
	CreateFunction(fCtx FunctionContext, args Arguments) (ExprFunc[K], error)
}

type CreateFunctionFunc func(fCtx FunctionContext, args Arguments) (ExprFunc[K], error)

type factory[K any] struct {
	name               string
	args               Arguments
	createFunctionFunc CreateFunctionFunc
}

func (f *factory[K]) Name() string {
	return f.name
}

func (f *factory[K]) CreateEmptyArguments() Arguments {
	return f.args
}

func (f *factory[K]) CreateFunction(fCtx FunctionContext, args Arguments) (ExprFunc[K], error) {
	return f.createFunctionFunc(fCtx, args)
}

// We'd use Options in an actual implementation probably
func NewFactory[K any](name string, args Arguments, createFunctionFunc CreateFunctionFunc) Factory[K] {
	f := &factory[K]{
		name:               name,
		args:               args,
		createFunctionFunc: createFunctionFunc,
	}
	return f
}

type DeleteKeyArguments[K any] struct {
	Target ottl.PMapGetter[K] `ottlarg:"0"`
	Key    string             `ottlarg:"1"`
}

func NewDeleteKeyFactory[K any]() ottl.Factory[K] {
	return ottl.NewFactory[K]("delete_key", &DeleteKeyArguments[K]{}, createDeleteKeyFunction[K])
}

func createDeleteKeyFunction[K any](fCtx ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
	args, ok := oArgs.(*DeleteKeyArguments[K])

	if !ok {
		return nil, fmt.Errorf("DeleteKeyFactory args must be of type *DeleteKeyArguments[K]")
	}

	return deleteKey(args.Target, args.Key), nil
}

func deleteKey[K any](target ottl.PMapGetter[K], key string) ottl.ExprFunc[K] {
	return func(ctx context.Context, tCtx K) (interface{}, error) {
		val, err := target.Get(ctx, tCtx)
		if err != nil {
			return nil, err
		}
		val.Remove(key)
		return nil, nil
	}
}

```

If I'm interpreting your example correctly you introduced `Expr[K] interface` so that each function wouldn't need to specify a `createFunction` to handle calling the actual business logic right?  If we stick with that interface to keep the boilerplate in the funcs themselve can we use a different name?  `Expr` is already a used term in OTTL.

Reply:
>Do you think the component factory pattern of using options to allow extending functionality offers anything that we won't be able to do with the current approach?

the problem with that approach is that it complicates a lot the logic to handle the optional on consumer side. Also we should be consistent across things in the collector world.

Reply:
Look at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18822/files#diff-c2e10bc775ef2385c57d8524ed140b8b90894faec2fe397dd96f2abc7bd47e30

this can become:

```golang
func NewDeleteKeyArgumentsFactory() ottl.Factory {
  return ottl.NewFacory("delete_key", &DeleteKeyArguments[K]{}, createFunction)
}

func createFunction(fCtx ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
	args, ok := oArgs.(*DeleteKeyArguments[K])

	if !ok {
		return nil, fmt.Errorf("DeleteKeyFactory args must be of type *DeleteKeyArguments[K]")
	}

	return deleteKey(args.Target, args.Key), nil
}
```

Reply:
> If I'm interpreting your example correctly you introduced Expr[K] interface so that each function wouldn't need to specify a createFunction to handle calling the actual business logic right? If we stick with that interface to keep the boilerplate in the funcs themselve can we use a different name? Expr is already a used term in OTTL.

I was playing around with replacing the `Expr` struct with an interface, but I would prefer to keep your solution that just uses `createFunction` and keeps things simpler. The one thing I had in mind with the `Expr` interface was keeping the possibility of stateful processing, but I think we can address that later if we decide its something we want to support.

> the problem with that approach is that it complicates a lot the logic to handle the optional on consumer side. Also we should be consistent across things in the collector world.

That makes sense, thank you for clarifying that.

Reply:
I was testing out a similar approach for a `NewFactory` function, but it came up short on a couple points that have been addressed here. I'll switch things over to try this out, thank you both for your input.

Reply:
👍